rotkehlchen/data/import_utils.py

```diff
@@ -0,0 +1,341 @@
+
+    @classmethod
+    def process_trades(cls, timestamp: Timestamp, data: List[BinanceCsvRow]) -> List[Trade]:
+        """Processes multirows data and stores it into Rotki's Trades
+        Each row has format: {'Operation': ..., 'Change': ..., 'Coin': ...}
+        Change is amount, Coin is asset
+        If amount is negative then this asset is sold, otherwise it's bought
+        """
+        # Because we can get mixed data (e.g. multiple Buys or Sells on a single timestamp) we need
+        # to group it somehow. We are doing it by grouping the highest bought with the highest
+        # sold value. We query usd equivalent for each amount because different Sells / Buys
+        # may use different assets.
+
+        # Querying price takes a long time, so we would like to avoid it,
+        # and therefore we check if all Buys / Sells use the same asset.
+        # If so, we can group by original amount.
+
+        # Checking assets
+        same_assets = True
+        assets: Dict[str, Optional[Asset]] = defaultdict(lambda: None)
+        for row in data:
+            if row['Operation'] == 'Fee':
+                cur_operation = 'Fee'
+            elif row['Change'] < 0:
+                cur_operation = 'Sold'
+            else:
+                cur_operation = 'Bought'
+            assets[cur_operation] = assets[cur_operation] or row['Coin']
+            if assets[cur_operation] != row['Coin']:
+                same_assets = False
+                break
+
+        # Querying usd value if needed
+        if not same_assets:
+            for row in data:
+                try:
+                    price = PriceHistorian.query_historical_price(
+                        from_asset=row['Coin'],
+                        to_asset=A_USD,
+                        timestamp=timestamp,
+                    )
+                except NoPriceForGivenTimestamp:
+                    # If we can't find price we can't group, so we quit the method
+                    log.warning(f'Couldn\'t find price of {row["Coin"]} on {timestamp}')
+                    return []
+                row['usd_value'] = row['Change'] * price
+
+        # Soring by amount
+        rows_by_operation = {}
+        for is_fee, rows in groupby(sorted(data, key=cls.is_fee_key), key=cls.is_fee_key):
+            rows_by_operation[is_fee] = sorted(
+                rows, key=lambda x: x['Change'] if same_assets else x['usd_value'],
+            )
+
+        # Grouping by combining the highest sold with the highest bought and the highest fee
+        # Using fee only we were provided with fee (checking by "True in rows_by_operation")
+        grouped_trade_rows = []
+        while len(rows_by_operation[False]) > 0:
+            cur_batch = [rows_by_operation[False].pop(0), rows_by_operation[False].pop(-1)]
+            if True in rows_by_operation:
+                cur_batch.append(rows_by_operation[True].pop(0))
```

Comment:
Ah, I thought that lists in python are clever and thus popping from the beginning takes `O(1)` time. Surely will change the approach to more optimal

Reply:
Btw, I pop from both sides of the list. So this optimization will make sense only for fee rows since I pop them from only one side. But still optimization, so will apply it.